p2p/server.go

```diff
@@ -693,6 +694,87 @@ func (srv *Server) setupListening() error {
 	return nil
 }
 
+// natMapLoop performs initialization mapping for nat and repeats refresh.
+func (srv *Server) natMapLoop(natm nat.Interface, protocol string, intport, extport int, name string, interval time.Duration) {
+	var (
+		internal   = intport
+		external   = extport
+		mapTimeout = interval
+
+		newLogger = func(p string, e int, i int, n nat.Interface) log.Logger {
+			return log.New("proto", p, "extport", e, "intport", i, "interface", n)
+		}
+	)
+
+	log := newLogger(protocol, external, internal, natm)
+
+	p, err := srv.NAT.AddMapping(protocol, intport, extport, name, mapTimeout)
+	if err != nil {
+		srv.log.Debug("Couldn't add port mapping", "err", err)
+	} else {
+		log.Info("Mapped network port")
+		if p != uint16(external) {
+			log.Debug("Already mapped port", extport, "use alternative port", p)
+			log = newLogger(protocol, int(p), internal, natm)
+			external = int(p)
+		}
+		// Set it directly because it is an operation that is performed
+		// before Server.run is executed.
+		switch protocol {
+		case "tcp":
+			srv.localnode.Set(enr.TCP(external))
+		case "udp":
+			srv.localnode.SetFallbackUDP(external)
+		}
+	}
+
+	refresh := time.NewTimer(mapTimeout)
+	defer func() {
+		refresh.Stop()
+		log.Debug("Deleting port mapping")
+		natm.DeleteMapping(protocol, external, internal)
+	}()
+
+	for {
+		select {
+		case _, ok := <-srv.quit:
+			if !ok {
+				return
+			}
+		case <-refresh.C:
+			log.Trace("Refreshing port mapping")
+			p, err := natm.AddMapping(protocol, external, internal, name, mapTimeout)
+			if err != nil {
+				log.Debug("Couldn't add port mapping", "err", err)
+			} else {
+				if p != uint16(external) {
+					log.Debug("Already mapped port", external, "use alternative port", p)
+					log = newLogger(protocol, int(p), internal, natm)
+					external = int(p)
+
+					if err := srv.changePort(protocol, uint16(external)); err != nil {
+						log.Debug("Couldn't change port on localnode", "err", err)
+					}
+				}
+			}
+			refresh.Reset(mapTimeout)
+		}
+	}
+}
+
+// changePort changes the port number of localnode.
+func (srv *Server) changePort(protocol string, port uint16) error {
```

Comment:
I still don't quite understand why you want to process port changes on the server main loop. You can just change the port on `LocalNode`, it is safe for concurrent use.

Reply:
I vaguely imagined something like `server.ChangePort()`, or rpc method `net_changePort`...
But I can't come up with a clear scenario.

As you said, I want to change it to the following. Is my understanding correct?
```go
// in Server nat refresh loop
for {
	select {
	case _, ok := <-srv.quit:
		if !ok {
			return
		}
	case <-refresh.C:
		log.Trace("Refreshing port mapping")
		p, err := natm.AddMapping(protocol, external, internal, name, mapTimeout)
		if err != nil {
			log.Debug("Couldn't add port mapping", "err", err)
		} else {
			if p != uint16(external) {
				log.Debug("Already mapped port", external, "use alternative port", p)
				log = newLogger(protocol, int(p), internal, natm)
				external = int(p)

				switch protocol {
				case "tcp":
					srv.localnode.Set(enr.TCP(external))
				case "udp":
					srv.localnode.SetFallbackUDP(external)
				}
			}
		}
		refresh.Reset(mapTimeout)
	}
}
```

Reply:
Yes that looks kind of what I wanted.

Reply:
Actually, the flow can be a bit simpler than what you have there. Whether or not `AddMapping` returns a port that is equal to the local port, we can just always call `Set` or `SetFallbackUDP` on `LocalNode`. Idempotent updates will simply be ignored.